src/zcap.rs

```diff
@@ -0,0 +1,473 @@
+                    (Some(ref id), URI::String(ref t_id)) => {
+                        // ensure proof target cap ID and given
+                        if id != t_id {
+                            result
+                                .errors
+                                .push("Target Capability IDs dont match".into())
+                        };
+                    }
+                    _ => result
+                        .errors
+                        .push("Missing proof target capability ID".into()),
+                };
+                // if there are invokers listed in the target, ensure the inoker here is the right one
+                if let (Some(URI::String(ref invoker)), Some(ref delegatee)) =
+                    (&target_capability.invoker, &proof.verification_method)
+                {
+                    if invoker != delegatee {
+                        result.errors.push("Incorrect Invoker".into());
+                    }
+                };
+                if result.errors.is_empty() {
+                    result.checks.push(Check::Proof);
+                };
+                result
+            }
+        }
+    }
+
+    // https://w3c-ccg.github.io/ld-proofs/
+    pub async fn generate_proof(
+        &self,
+        jwk: &JWK,
+        options: &LinkedDataProofOptions,
+        target: &URI,
+    ) -> Result<Proof, Error> {
+        let cl = options.clone();
+        let opts = LinkedDataProofOptions {
+            property_set: match (cl.property_set, target) {
+                (Some(mut ps), URI::String(t)) => {
+                    ps.insert("capability".into(), Value::String(t.to_string()));
+                    Some(ps)
+                }
+                (_, URI::String(t)) => {
+                    let mut ps = Map::<String, Value>::new();
+                    ps.insert("capability".into(), Value::String(t.to_string()));
+                    Some(ps)
+                }
+            },
+            ..cl
+        };
+        LinkedDataProofs::sign(self, &opts, jwk).await
+    }
+
+    /// Prepare to generate a linked data proof. Returns the signing input for the caller to sign
+    /// and then pass to [`ProofPreparation::complete`] to complete the proof.
+    pub async fn prepare_proof(
+        &self,
+        public_key: &JWK,
+        options: &LinkedDataProofOptions,
+        target: &URI,
+    ) -> Result<ProofPreparation, Error> {
+        let mut prep = LinkedDataProofs::prepare(self, options, public_key).await?;
+        prep.proof.property_set = match (prep.proof.property_set, target) {
+            (Some(mut ps), URI::String(t)) => {
+                ps.insert("capability".into(), Value::String(t.to_string()));
+                Some(ps)
+            }
+            (_, URI::String(t)) => {
+                let mut ps = Map::<String, Value>::new();
+                ps.insert("capability".into(), Value::String(t.to_string()));
+                Some(ps)
+            }
+        };
+        Ok(prep)
```

Comment:
The proof and signing input in `prep` must match. The signing input may be a digest of the proof and linked data document (e.g. invocation). So `capability` should be added to `options` before calling `LinkedDataProofs::prepare`, to ensure it is included in the signing input. `prepare` functions in `ldp.rs` will need to be updated to pass through the `capability` option, and/or additional options in general.

Reply:
:point_up: related: https://github.com/spruceid/ssi/commit/41dbff3bbef37d29a9da1fa359503d8f869a5a97

Reply:
updated to match `::generate_proof`. I experimented with generic signing inputs on branch `feat/generic_proof`, but the refactor got quite large and it wasnt my primary goal

Reply:
73e4ac1e34f1e3ba00f33f5c686c81492b97a8bb; good, `LinkedDataProofs::sign` also has to be updated to pass through `capability` and/or `property_set` from the LinkedDataProofOptions object into the Proof object (i.e. at each occurence of `proof = Proof {` in `src/ldp.rs`)